internal/component/discovery/discovery.go

```diff
@@ -216,8 +216,9 @@ func (c *Component) runDiscovery(ctx context.Context, d DiscovererWithMetrics) {
 	}
 
 	ticker := time.NewTicker(MaxUpdateFrequency)
-	// true if we have received new targets and need to send.
-	haveUpdates := false
+	// true if we have received new targets and need to send. Initially set it to true to send empty targets in case
+	// the discoverer never sends any targets.
+	haveUpdates := true
```

Comment:
Will it send empty targets if there are already targets in the cache?

Reply:
If targets are added to `cache` map, this means that the `case groups := <-ch:` triggered before `case <-ticker.C:`. The `haveUpdates` remains `true`. 

Then the `case <-ticker.C:` would typically trigger and we would publish the targets in the `cache` map. 

However, if the underlying discovery never publishes anything, the `case groups := <-ch:` will not be hit and instead the `case <-ticker.C:` would be triggered first and an empty `cache` map would be published. 

Hope this answers your question!

Reply:
Yes, that's right! I forgot that the cache is local to this function. It makes sense now.